### PR TITLE
Ghseeli/specialized catalan functions

### PR DESCRIPTION
**Pull request by ghseeli**
I added the interface to specialize `t` in the evaluation method of a `CatalanFunction`. This seemed like the best place to put it without modifying any of the existing code too heavily.

### CODE REVIEW

**Comment by MareoRaft** [comment]
Can you add some tests to `test_all.py` showing how you would plug in t=1 when creating a catalan function?  Make sure all tests in `test_all.py` continue to pass.

**Comment by MareoRaft** [comment]
The tests normally pass in under 6 seconds.  

    Sym = SymmetricFunctions(QQ['t'])
    hl = Sym.hall_littlewood().Qp()
    p = Partition([5, 5, 3, 3, 1, 1])
    a(compositional_hall_littlewood_Qp(p), hl(p))

is the slow point.  I think something in your changes causes a slow down.

**Comment by MareoRaft** [comment]
The slow down ultimately comes from passing `t` into `SymmetricFunctions(self.base_ring).hall_littlewood(t=t).Qp()`.  Using `SymmetricFunctions(self.base_ring).hall_littlewood().Qp()` is significantly faster.

**Comment by MareoRaft** [comment]
The only thing that I don't understand is your change of `compositional_hall_littlewood_Qp(gamma, t=s.t, base_ring=s.base_ring())`.  Do we need that change?

**Comment by ghseeli** [comment]
Re the `repr` comment, the problem is that the code does not specify `t` in the `init` method, it instead gets evaluated at the `eval()` method. This was the least work way to modify the catalan functions. 

Odd that it causes such a slowdown, I can change it up slightly.

I put tests in the docstring, but I can add them to `test_all.py`, too.

**Comment by MareoRaft** [comment]
As for issue #9 , it seems that the line `return H(gamma)(HLQp.one())` was the culprit.  I forgot to pass in the base ring.  Now it's `return H(gamma, base_ring=base_ring)(HLQp.one())`.

Pushing my version up shortly.

File: k_combinat_for_sage/all.py [hunk 1]
```diff
@@ -170,30 +170,30 @@ def number_of_noninversions(lis):
                     num += 1
         return num
     gamma = Composition(gamma)
-    if s.__class__.__name__ in ('SymmetricFunctionAlgebra_monomial_with_category', 'SymmetricFunctionAlgebra_dual_with_category'):
+    if basis.__class__.__name__ in ('SymmetricFunctionAlgebra_monomial_with_category', 'SymmetricFunctionAlgebra_dual_with_category'):
         raise NotImplemented(
             'Straightening does not exist (that i know of) for the monomial basis or the forgotten/dual basis.')
-    elif s.__class__.__name__ == 'HallLittlewood_qp_with_category':
-        return compositional_hall_littlewood_Qp(gamma, t=s.t, base_ring=s.base_ring())
```
Comment (ghseeli) [inline k_combinat_for_sage/all.py]: Why did you delete this line? It is very important. Otherwise, there will be hidden t's in the HLQp terms and the specialization does not work correctly.

**Comment by ghseeli** [comment]
For more information on my in-line comment, the compositional hall-littlewood needs to be passed a t because specializing t affects both the raising operator expression AND what they act on (using the hall-littlewood definition).

File: k_combinat_for_sage/all.py (hunk 1 above)
Comment (MareoRaft) [reply k_combinat_for_sage/all.py]: No reason other than I couldn't understand why you needed it, and I needed to troubleshoot.  I think I understand now.  When you create the parent basis `hall_littlewood(t=val)`, it sets the attribute `.t` equal to val, so here you are making sure that value gets passed along when the new straightened version of the hall littlewood is made.  Correct me if I'm wrong.  You can add this back in with your other changes.

Comment (ghseeli) [reply k_combinat_for_sage/all.py]: Yeah, that is correct!

**Review by MareoRaft — APPROVED** [review]

**Comment by MareoRaft** [comment]
I think passing in `t` in `return compositional_hall_littlewood_Qp(gamma, base_ring=basis.base_ring(), t=basis.t)` may slow down things in certain situations.  Hopefully won't become a problem though

**Comment by MareoRaft** [comment]
Thanks, George!!
